KAFKA-3282: Change tools to use --new-consumer by default and introduce --old-consumer

arunmahadevan opened the pull request:
1. Made new consumer the default in console consumer
2. Added --old-consumer option to choose old consumer

arunmahadevan committed:
KAFKA-3706: Make new consumer as default in kafka console consumer

1. Made new consumer the default in console consumer
2. Added --old-consumer option to choose old consumer

arunmahadevan commented:
Right now addresses only console consumer. Will update other tools as well.

harshach commented:
+1

arunmahadevan committed:
KAFKA-3706: Add old-consumer option in ConsumerGroupCommand and ConsumerPerformance

arunmahadevan commented:
So far addressed the following 
- ConsoleConsumer
- ConsumerGroupCommand
- ConsumerPerformance

gwenshap commented:
Hey, @harshach - any reason you didn't merge this?

ewencp commented:
It doesn't seem safe to add the `--old-consumer` option and change the default in one release. How can people migrate their scripts if there isn't a window during which both `--old-consumer` and `--new-consumer` options work? Seems like it could catch people unawares with a new release where they will have had to provide paths for both cases to support a seamless migration. Something like `ConsumerPerformance` is probably low risk, `ConsumerGroupCommand` seems like it could break things for a lot of people... (Ideally we'd have a way besides release notes to also warn of the switch; something like deprecation but with a slightly different meaning.)

arunmahadevan commented:
@ewencp the proposed patch doesn't remove the `--new-consumer` option. Both `--old-consumer` and `--new-consumer` are supported, with new consumer being the default if neither options are provided. This would probably break the user scripts that uses the current default (old consumer) since the new consumer expects a `bootstrap-server` instead of `zookeeper url`. They would have to either add an extra `--old-consumer` option if they wish to continue with the old consumer or move to the new consumer.

May be we can show a warning when `--old-consumer` option is used and prompt users to migrate. If no option is provided but zookeeper url is present (current default), it would throw the same warning and just exit.

ewencp commented:
@arunmahadevan Yeah, the second sentence was unclear, but my first sentence:

> It doesn't seem safe to add the --old-consumer option and change the default in one release.

Was what I meant. Normally the way stuff like this is handled is to ensure the new option is made available in one release without changing defaults so that scripts don't break but you can warn users that there will be a change, then change the default in a subsequent release. You can try to make both work based on required arguments, though that makes things a bit messier (and in which case the `--new-consumer` and `--old-consumer` arguments are actually superfluous and only serve to make the decision explicit since they have other mutually exclusive required arguments).

arunmahadevan commented:
@ewencp thanks for the clarification.So the options are,
1. Keep the current default (old consumer) and just display a warning message that the new consumer would be made default in a subsequent release.
2. Make new consumer as default and support both --new-consumer and --old-consumer. As a workaround to not break existing scripts, switch to old consumer if `zookeeper url` is provided (current default option) and display the warning message.

Option 2 may be better since the new consumer can be made default without breaking existing scripts. If it sounds ok I can make the necessary changes.

ewencp commented:
@arunmahadevan For option 2, I think you need to be very careful. If you go that route, it seems like the `--new-consumer` and `--old-consumer` flags don't really do anything beyond having `--zookeeper` and `--bootstrap-server` opts. You could "default" to new consumer in the sense that if `--zookeeper` is not provided, any error messages would say that the new consumer parameters are required. This would guide new users of the command to the new consumer version, although clearly some users will still need the old consumer versions if they still have old consumer groups.

What's unclear to me is whether including the `--old-consumer` flag makes much sense since a `--zookeeper` is required in that case anyway; it just makes the decision explicit and can help guide the user re: required arguments if the help message explains to select via `--old-consumer` and `--new-consumer` first, then gives help customized to each case.

arunmahadevan committed:
KAFKA-3706: Remove --old-consumer option

Select old-consumer based on zk option and display a warning
message prompting to migrate.

arunmahadevan commented:
@ewencp I've removed the `--old-consumer` option and made new consumer as the default. The commands will use old consumer if the zookeeper option is specified (so that the existing scripts will continue to run). A message is displayed to prompt the user to migrate in this case.

 `--new-consumer` option is retained for backward compatibility.

ijuma commented:
It's not clear to me why we are deprecating the old consumer option since the old consumer itself is not deprecated yet. It seems appealing to deprecate both at the same time and the old consumer won't be deprecated until after 0.10.1.0. I don't feel strongly either way though.

We had a recent discussion in the mailing list about removing the beta label from the new consumer and people were in favour. As such, it would be great to include this PR in the next release. The minimum set of changes IMO are:
1. Make the new consumer the default as long as `--zookeeper` is not specified
2. Fail with an error if `--zookeeper` and `--new-consumer` are specified at the same time
3. Support the `--new-consumer` option for compatibility
4. Update usages of the tools in the docs to use the new consumer
5. Add a note in upgrade.html

I am not sure if we should support `--old-consumer`. The main reason would be to be consistent with the console producer, but probably not worth it.

@arunmahadevan, can you please rebase the PR?

arunmahadevan commented:
@ijuma rebased with trunk.

ijuma commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -54,8 +54,12 @@ object ConsumerGroupCommand {
-      if (opts.options.has(opts.newConsumerOpt)) new KafkaConsumerGroupService(opts)
-      else new ZkConsumerGroupService(opts)
+      if (opts.useOldConsumer) {
+        Console.err.println("Old consumer is deprecated and will be removed in a future release. Consider using the new consumer.")
```
We should not deprecate this because the old consumer is not deprecated and if they are using ZK-based consumer groups, this is the only way to get the required information.

arunmahadevan replied:
Changed to `Using the ConsumerGroupCommand with old consumer is deprecated and will be removed in a future release. Consider using the new consumer.`. Is this ok?

ijuma replied:
I don't think we should do this because there are good reasons to use this command with the old consumer (i.e. ZK-based offsets).

ijuma commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -420,27 +424,28 @@ object ConsumerGroupCommand {
       // check required args
-      if (options.has(newConsumerOpt)) {
+      if (useOldConsumer) {
+        CommandLineUtils.checkRequiredArgs(parser, options, zkConnectOpt)
```
We are checking `zkConnectOpt` here, but we also check it in `useOldConsumer`.

ijuma commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -420,27 +424,28 @@ object ConsumerGroupCommand {
+        CommandLineUtils.checkRequiredArgs(parser, options, zkConnectOpt)
+
+        if (options.has(bootstrapServerOpt))
+          CommandLineUtils.printUsageAndDie(parser, s"Option $bootstrapServerOpt is not valid with old consumer.")
```
This message may be confusing because users have not specified `old consumer`, they just specified `zookeeper`. There are a few other cases like this.

ijuma commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -420,27 +424,28 @@ object ConsumerGroupCommand {
 
         if (options.has(zkConnectOpt))
-          CommandLineUtils.printUsageAndDie(parser, s"Option $zkConnectOpt is not valid with $newConsumerOpt")
+          CommandLineUtils.printUsageAndDie(parser, s"Option $zkConnectOpt is only valid with old consumer.")
```
This can't happen because the `else` only runs if `zkConnectOpt` is not present.

ijuma commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -361,7 +362,7 @@ object ConsoleConsumer extends Logging {
       else OffsetRequest.LatestTime
 
-    CommandLineUtils.checkRequiredArgs(parser, options, if (useNewConsumer) bootstrapServerOpt else zkConnectOpt)
+    CommandLineUtils.checkRequiredArgs(parser, options, if (useOldConsumer) zkConnectOpt else bootstrapServerOpt)
```
We don't need to check `zkConnectOpt` as that is how `useOldConsumer` is defined.

ijuma commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -314,21 +314,22 @@ object ConsoleConsumer extends Logging {
         CommandLineUtils.printUsageAndDie(parser, "Exactly one of whitelist/blacklist/topic is required.")
       topicArg = options.valueOf(topicOrFilterOpt.head)
       filterSpec = if (options.has(blacklistOpt)) new Blacklist(topicArg) else new Whitelist(topicArg)
+      Console.err.println("Old consumer is deprecated and will be removed in a future release. Consider using the new consumer.")
```
I think we should say that using the ConsoleConsumer with the old consumer is deprecated. The old consumers are not deprecated themselves yet (but they will be in the release after 0.10.1.0).

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -50,6 +50,9 @@ object ConsumerPerformance {
     val consumerTimeout = new AtomicBoolean(false)
 
+    if (config.useOldConsumer)
+      Console.err.println("Old consumer is deprecated and will be removed in a future release. Consider using the new consumer.")
```
We should not deprecate this as it's useful to compare the performance of the old and new consumers.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -89,7 +89,7 @@ class ConsoleConsumerTest extends JUnitSuite {
 
     //Then
-    assertFalse(config.useNewConsumer)
+    assertTrue(config.useOldConsumer)
```
Doesn't the old assertion work any more? We should probably keep it to ensure that we are compatible.

arunmahadevan replied:
The `useNewConsumer` boolean flag is removed.  `useOldConsumer` flag is used instead so the assertion is just inverted.

ijuma replied:
OK.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -108,14 +108,14 @@ class ConsoleConsumerTest extends JUnitSuite {
 
     //Then
-    assertTrue(config.useNewConsumer)
+    assertFalse(config.useOldConsumer)
```
Doesn't the old assertion work any more? We should probably keep it to ensure that we are compatible.

arunmahadevan replied:
The `useNewConsumer` boolean flag is removed.  `useOldConsumer` flag is used instead so the assertion is just inverted.

ijuma replied:
OK. Not sure what happened and why GitHub repeated my question several times. :) I deleted the other ones.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -128,12 +128,28 @@ class ConsoleConsumerTest extends JUnitSuite {
 
     //Then
-    assertTrue(config.useNewConsumer)
+    assertFalse(config.useOldConsumer)
```
Doesn't the old assertion work any more? We should probably keep it to ensure that we are compatible.

arunmahadevan replied:
The `useNewConsumer` boolean flag is removed.  `useOldConsumer` flag is used instead so the assertion is just inverted.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -150,7 +166,7 @@ class ConsoleConsumerTest extends JUnitSuite {
 
     //Then
-    assertTrue(config.useNewConsumer)
+    assertFalse(config.useOldConsumer)
```
Doesn't the old assertion work any more? We should probably keep it to ensure that we are compatible.

arunmahadevan replied:
The `useNewConsumer` boolean flag is removed.  `useOldConsumer` flag is used instead so the assertion is just inverted.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -165,7 +181,7 @@ class ConsoleConsumerTest extends JUnitSuite {
     propsStream.write("consumer.timeout.ms=1000".getBytes())
```
Maybe we want to update this config to use `request.timeout.ms` which is a valid config for the new consumer.

arunmahadevan replied:
done

ijuma reviewed (changes requested):
Thanks for the PR, I left a few comments.

arunmahadevan committed:
Addressed review comments from @ijuma

ijuma commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -54,8 +54,13 @@ object ConsumerGroupCommand {
-      else new ZkConsumerGroupService(opts)
+      if (opts.useOldConsumer) {
+        Console.err.println("Using the ConsumerGroupCommand with old consumer is deprecated and will be removed in a future release. " +
+          "Consider using the new consumer.")
```
I don't think we should deprecate this. As I said in my comment: "We should not deprecate this because the old consumer is not deprecated and if they are using ZK-based consumer groups, this is the only way to get the required information."

ijuma commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -420,27 +425,22 @@ object ConsumerGroupCommand {
       // check required args
-      if (options.has(newConsumerOpt)) {
+      if (useOldConsumer) {
+        if (options.has(bootstrapServerOpt))
```
We should probably throw an error if --new-consumer is specified too.

ijuma commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -314,21 +314,23 @@ object ConsoleConsumer extends Logging {
         CommandLineUtils.printUsageAndDie(parser, "Exactly one of whitelist/blacklist/topic is required.")
       topicArg = options.valueOf(topicOrFilterOpt.head)
       filterSpec = if (options.has(blacklistOpt)) new Blacklist(topicArg) else new Whitelist(topicArg)
+      Console.err.println("Using the ConsoleConsumer with old consumer is deprecated and will be removed " +
```
We should probably check that `--new-consumer` has not been specified together with `zkConnect`

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -213,13 +213,14 @@ object ConsumerPerformance {
     else
       new Properties
-    if (useNewConsumer) {
+    if(!useOldConsumer) {
```
Space after `if`

ijuma replied:
Also, we should probably check that `zkConnect` is not used together with `--new-consumer`.

arunmahadevan committed:
addresed a few more comments

arunmahadevan commented:
@ijuma thanks for the detailed review. I 've addressed the latest comments, can you take a look again?

ijuma commented:
@arunmahadevan, thanks. I have to test this still, but the code changes look good to me. 

@hachikuji, do you think we can still include this in the 0.10.1 branch? It seems low risk as current users of the old consumer will continue to use the old consumer because of the `--zookeeper` parameter.

The main changes are that users can use the new consumer without having to specify `--new-consumer` and we note to console consumer users that the option of using the old consumer with it is deprecated and will be removed in a future release. I started a system tests run here to verify that this doesn't cause issues: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/

hachikuji commented:
@ijuma Anything to decrease usage of the old consumer! One thing we should probably be a little careful about is all the system tests which depend on the console consumer (though the `ConsoleConsumer` test service). I think the main usage is in the `ProduceConsumeValidateTest`, but each test which extends it uses different options to construct the consumer. In most cases, I think we we want the test cases to be using the new consumer, but we should make sure the old consumer is still covered.

ijuma commented:
@hachikuji Yeah, I think the existing system tests should continue with their existing behaviour since they do:

``` python
if self.new_consumer:
            cmd += " --new-consumer --bootstrap-server %(broker_list)s" % args
        else:
            cmd += " --zookeeper %(zk_connect)s" % args
```

I started the system tests build to double-check.

Like you, I was thinking that we should change their behaviour so that the new consumer is used by default while ensuring that the old consumer is still covered. I filed https://issues.apache.org/jira/browse/KAFKA-4211 to track that for 0.10.2.0.

hachikuji commented:
@ijuma Makes sense. Thanks for the explanation.

hachikuji commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -128,12 +128,28 @@ class ConsoleConsumerTest extends JUnitSuite {
+
+  @Test
+  def testDefaultConsumer() {
+    //Given
```
nit: Not a big deal, but these comments are a little annoying. The tests seem clear enough without them.

ijuma replied:
I think @arunmahadevan was following the pattern in the other tests in the class. I agree with you though.

hachikuji replied:
Yep, I noticed. I wouldn't hate it if we removed them all 😄 , but the patch shouldn't block on it.

ijuma commented:
The initial system tests run had 3 unrelated failures. I rebased the branch on trunk and the system tests then passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/98/

I also manually tested the 3 command-line tools changed in this PR.

LGTM, merged to trunk and 0.10.1 branches after a minor tweak of the deprecation message we output for the console consumer.
